packages/evm/contracts/Yaru.sol

```diff
@@ -1,70 +1,70 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 pragma solidity ^0.8.17;
 
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import { IYaru } from "./interfaces/IYaru.sol";
 import { IHashi, IOracleAdapter } from "./interfaces/IHashi.sol";
 import { Message } from "./interfaces/IMessage.sol";
-import { IMessageExecutor } from "./interfaces/IMessageExecutor.sol";
 import { MessageHashCalculator } from "./utils/MessageHashCalculator.sol";
+import { IJushinki } from "./interfaces/IJushinki.sol";
 
-contract Yaru is IMessageExecutor, MessageHashCalculator {
-    IHashi public immutable hashi;
-    address public immutable yaho;
-    uint256 public immutable chainId;
-    mapping(uint256 => bool) public executed;
-    address public sender;
+contract Yaru is IYaru, MessageHashCalculator, ReentrancyGuard, Ownable {
+    address public immutable hashi;
+    address public immutable headerVault;
```

Comment:
`Yaru` should be use-case agnostic, so it should refer to "hashes" rather than "headers".

I'd prefer that there not be something like `HeaderVault` enshrined at the `Yaru` layer. It should be a layer built on top, as per the above comment about `Ownable`.

It seems to me that this functionality belongs in a contract in the `ownable` folder.

Reply:
Let me explain. 

With the introduction of EIP-5164, the old messageId calculation has become obsolete. This is because using block numbers as if they were message IDs is not compliant, and there is also the possibility of having messages with the same message ID. For this reason, we have decided to always use `Yaho` for sending messages, whether it's for block header reporting or normal message passing.

Regarding the reason why we decided to add a link to `HeaderVault` within `Yaru`, it is because we need a way to natively distinguish it from normal contracts.

This approach greatly simplifies the flow. We will essentially have a unique process for sending both blocks and messages. External oracles will only need to write the `MessageRelay` and Adapter contracts, eliminating the need for a `HeaderReporter`. Moreover, with this approach, it is always possible to fetch a block hash from `Hashi` given the block number, ensuring compatibility with the standard. Ex:

```
function getBlockHeadersFromOracles(
        IOracleAdapter[] calldata oracles,
        uint256 fromChainId,
        uint256 blockNumber
    ) external view returns (bytes32[] memory) {
        return
            IHashi(hashi).getHashesFromOracles(
                oracles,
                fromChainId,
                calculateMessageId(
                    fromChainId,
                    block.chainid,
                    IYaru(yaru).yahos(fromChainId),
                    address(0),
                    address(0),
                    keccak256(abi.encode(blockNumber)),
                    MESSAGE_BHR
                )
            );
    }
```

Reply:
Wait... why relay blocks through `Yaru`? That seems needlessly complicated?
The whole point of Hashi is to be modular. `Yaru` can be EIP-5164 compliant, and be used for passing messages (which may contain block headers in some cases), but there is no reason that all routes should be forced to use `Yaru`.

Reply:
> there is also the possibility of having messages with the same message ID

This is not an issue, since it's up to the respective oracle adapters to determine what IDs mean.

Reply:
it's not really like that! let's say that the blocks might not even pass through `Yaru` since you would still have the hash saved in each adapter. The idea of ​​the `HeaderVault` was to natively manage the execution of a message containing a block header. A possible solution is to reverse the execution of those messages.

Reply:
> A possible solution is to reverse the execution of those messages.

Not sure what you mean here. Could you elaborate some?

> The idea of ​​the HeaderVault was to natively manage the execution of a message containing a block header.

This is exactly my point though. I don't think this should be native to `Yaru`, rather it should be a separate contract that `Yaru` passes a message to.

Reply:
Hashi is constructed as a collection of distinct and useful "stages". This is a really valuable feature from the point of view of composability. Wherever possible, we should resist the urge to bundle things together / natively support specific use-cases and the lower/more generalized stages.

Reply:
Since now block header messages and normal ones have the same format (same `Message` struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages. 

I understand your point of building `HeaderVault` on top of `Yaru` but in that case, however, the propagation of the blocks would no longer be native because it could be done by everyone.


> > A possible solution is to reverse the execution of those messages.
> 
> Not sure what you mean here. Could you elaborate some?
> 
> > The idea of ​​the HeaderVault was to natively manage the execution of a message containing a block header.
> 
> This is exactly my point though. I don't think this should be native to `Yaru`, rather it should be a separate contract that `Yaru` passes a message to.

Reply:
> the propagation of the blocks would no longer be native because it could be done by everyone.

I don't see why this is the case. The HeaderVault would simply validate the sending address.

> Since now block header messages and normal ones have the same format (same Message struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages.

If `HeaderVault` is a separate contract, then the "execution" of a message is simply the call to `HeaderVault`. In other words, there is no need for any special handling in this case.

Reply:
This is functionally similar to the existing `HeaderStorage` contract.

Reply:
> > the propagation of the blocks would no longer be native because it could be done by everyone.
> 
> I don't see why this is the case. The HeaderVault would simply validate the sending address.
> 
> > Since now block header messages and normal ones have the same format (same Message struct) they could be both potentially executed so if we don't want to put the `HeaderVault` address within `Yaru` we have to block the execution of block header messages.
> 
> If `HeaderVault` is a separate contract, then the "execution" of a message is simply the call to `HeaderVault`. In other words, there is no need for any special handling in this case.

Correct! But then in order to access the block header hash within `Hashi` you would need also the `message.from` and the `message.to`, which is not intuitive and as far as i know this was a requirement.

Reply:
This is the job of the adapters though.
Like if my adapter is using Yaru, then it's my adapter's job to know who the sender should be.